Permission WRITE_EXTERNAL_STORAGE missing

### Describe your problem!

Trying to reproduce some other bug on my physical testing device, c:geo asked me whether to update my routing data. On selecting to update it, c:geo crashed. Logcat (see below) showed a missing WRITE_EXTERNAL_STORAGE permission.

### How to reproduce?

For me:
- uninstalled existing release 2023.02.13
- installed dev build from current state of `release`
- followed wizard, granting location permission, selecting base folder, and then restoring backup
- tapped on home screen notification whether to update downloaded routing data

### Actual result after these steps?

crash

Logcat shows this:
```
FATAL EXCEPTION: network--4
Process: cgeo.geocaching, PID: 12409
io.reactivex.rxjava3.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | java.lang.SecurityException: No permission to write to /storage/emulated/0/Download/E5_N50.rd5: Neither user 10285 nor current process has android.permission.WRITE_EXTERNAL_STORAGE.
	at io.reactivex.rxjava3.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:372)
	at io.reactivex.rxjava3.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:45)
	at io.reactivex.rxjava3.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:25)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
	at java.lang.Thread.run(Thread.java:818)
Caused by: java.lang.SecurityException: No permission to write to /storage/emulated/0/Download/E5_N50.rd5: Neither user 10285 nor current process has android.permission.WRITE_EXTERNAL_STORAGE.
	at android.os.Parcel.readException(Parcel.java:1620)
	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:183)
	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:135)
	at android.content.ContentProviderProxy.insert(ContentProviderNative.java:476)
	at android.content.ContentResolver.insert(ContentResolver.java:1284)
	at android.app.DownloadManager.enqueue(DownloadManager.java:1336)
	at cgeo.geocaching.downloader.DownloaderUtils.lambda$triggerDownloads$3(DownloaderUtils.java:202)
	at cgeo.geocaching.downloader.DownloaderUtils$$ExternalSyntheticLambda4.run(D8$$SyntheticClass)
	at io.reactivex.rxjava3.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38)
```

received a similar error on next restart, when I tried to update map data

### Expected result after these steps?

check existing routing data and display update dialog

### Reproducible

Yes

### c:geo Version

release as of 2023-02-17

### System information

```text
## System information

c:geo version: 2023.02.17-06ae7c6 developer build

Device:- Device type: SM-G903F (s5neoltexx, samsung)
- Available processors: 8
- Android version: 6.0.1
- Android build: MMB29K.G903FXXU2BRH2
- Screen resolution: 1080x1920px (360x616dp)
- Pixel density: 3.0
- System font scale: 1.0 / used scale: 1.0
- Sailfish OS detected: false
- Google Play services: enabled - 22.50.14 (040308-499278674)

Sensor and location:- Low power mode: inactive
- Compass capabilities: yes
- Rotation vector sensor: present
- Orientation sensor: present
- Magnetometer & Accelerometer sensor: present
- Direction sensor used: rotation vector

Program settings:- Settings: v9, Count:80
- Set language: de_DE (system default)
- System date format: dd.MM.yy
- Time zone: GMT+01:00
- Debug mode active: no
- Last backup: 12. Feb., 22:44
- Routing mode: Walk
- Live map mode: true
- OSM multi-threading: false / threads: 1
- Map: Google: Karte
  - Id: cgeo.geocaching.maps.google.v2.GoogleMapProvider$GoogleMapSource
  - Atts: none
  - Theme: none

Filters:- Hide waypoints: -
- LIVE: Keine ([:advanced=false:inconclusive=false])
- OFFLINE: Keine ([:advanced=false:inconclusive=false])

Stored Filters (#0):

Services:- Geocaching sites enabled:
   geocaching.com: Logged in (Anmeldung OK) / BASIC
- Geocaching.com date format: M/d/yyyy
- Geocaching.com website language: en-US
- Last successful login on geocaching.com: 2023-02-17 20-32
- Routing: internal / BRouter installed: false
- Installed c:geo plugins: contacts

Permissions-android.permission.ACCESS_COARSE_LOCATION:: granted
-android.permission.ACCESS_FINE_LOCATION:: granted
-android.permission.READ_CONTACTS:: DENIED
-android.permission.READ_EXTERNAL_STORAGE:: DENIED

Paths- System internal c:geo dir: /data/user/0/cgeo.geocaching (1,6 GB free) v3 internal isDir(9 entries)
- Legacy User storage c:geo dir: /data/user/0/cgeo.geocaching (1,6 GB free) v3 internal isDir(9 entries)
- Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (1,5 GB free) v3 external non-removable isDir(2 entries)
- Internal theme sync (is turned off): /data/user/0/cgeo.geocaching/MapThemeData (1,6 GB free) v3 internal isDir(0 entries)
- Public Folders: #12
  - BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=49, dirs:>=10, totalFileSize:>=671,7 MB, free space: 1,5 GB, files on device: 369824)
  - OFFLINE_MAPS: /cgeo/maps (Default)[/cgeo/maps[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/maps]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fmaps, Av:true, files:7, dirs:2, totalFileSize:667,4 MB, free space: 1,5 GB, files on device: 369824)
  - OFFLINE_MAP_THEMES: /cgeo/maps/_themes (Default)[/cgeo/maps/_themes[PERSISTABLE_FOLDER(OFFLINE_MAPS)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/maps/_themes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fmaps%2F_themes, Av:true, files:2, dirs:0, totalFileSize:518,8 KB, free space: 1,5 GB, files on device: 369824)
  - OFFLINE_MAP_SHADING: /cgeo/maps/_hgt (Default)[/cgeo/maps/_hgt[PERSISTABLE_FOLDER(OFFLINE_MAPS)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/maps/_hgt]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fmaps%2F_hgt, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 1,5 GB, files on device: 369824)
  - LOGFILES: /cgeo/logfiles (Default)[/cgeo/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/logfiles]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Flogfiles, Av:true, files:8, dirs:0, totalFileSize:658,8 KB, free space: 1,5 GB, files on device: 369824)
  - GPX: /cgeo/gpx (Default)[/cgeo/gpx[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/gpx]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fgpx, Av:true, files:34, dirs:0, totalFileSize:3,7 MB, free space: 1,5 GB, files on device: 369824)
  - BACKUP: /cgeo/backup (Default)[/cgeo/backup[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/backup]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fbackup, Av:true, files:>=8, dirs:>=11, totalFileSize:>=6,6 MB, free space: 1,5 GB, files on device: 369824)
  - FIELD_NOTES: /cgeo/field-notes (Default)[/cgeo/field-notes[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/field-notes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Ffield-notes, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 1,5 GB, files on device: 369824)
  - SPOILER_IMAGES: /cgeo/GeocachePhotos (Default)[/cgeo/GeocachePhotos[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/GeocachePhotos]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2FGeocachePhotos, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 1,5 GB, files on device: 369824)
  - ROUTING_BASE: /cgeo/routing (Default)[/cgeo/routing[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting, Av:true, files:9, dirs:1, totalFileSize:229,1 MB, free space: 1,5 GB, files on device: 369824)
  - ROUTING_TILES: /cgeo/routing/segments4 (Default)[/cgeo/routing/segments4[PERSISTABLE_FOLDER(ROUTING_BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing/segments4]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting%2Fsegments4, Av:true, files:2, dirs:0, totalFileSize:229,0 MB, free space: 1,5 GB, files on device: 369824)
  - TEST_FOLDER: [Legacy]/data/user/0/cgeo.geocaching/files/unittest (Default)[/data/user/0/cgeo.geocaching/files/unittest[FILE#1:p-file:///data/user/0/cgeo.geocaching/files::/unittest]] (Uri: file:///data/user/0/cgeo.geocaching/files/unittest, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 1,5 GB, files on device: -1)
- Map render theme path:
- PersistedDocumentUris: #1
- UNUSED_TRACK: null
- Persisted Uri Permissions: #1
  - content://com.android.externalstorage.documents/tree/primary%3Acgeo (17. Feb., 20:22):RW
- Database: /data/user/0/cgeo.geocaching/databases/data (v99, Size:236,0 KB) on system internal storage

--- End of system information ---
```


### Additional Information

_No response_

Started testing with emulators:

API 23 (Android 6)
- open map
- select "check for missing routing data" => some missing entries listed
- select "ok" => crash

`Caused by: java.lang.SecurityException: No permission to write to /storage/emulated/0/Download/E5_N45.rd5: Neither user 10061 nor current process has android.permission.WRITE_EXTERNAL_STORAGE.`

So it looks rather version-specific than vendor- or device-specific.

Interestingly the system is blocking writing to the `Download` folder, so not SAF-specific c:geo folder access.

Will continue by creating emulators for newer Android versions.

--- a/main/AndroidManifest.xml
+++ b/main/AndroidManifest.xml
@@ -14,6 +14,7 @@
     <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
     <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
     <uses-permission android:name="android.permission.NFC" />
+    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" android:maxSdkVersion="29" />
     <!-- <uses-permission android:name="android.permission.READ_CONTACTS" /> Activate when SDKmin=23, see #13777 -->
 
     <uses-feature

--- a/main/src/main/java/cgeo/geocaching/InstallWizardActivity.java
+++ b/main/src/main/java/cgeo/geocaching/InstallWizardActivity.java
@@ -40,6 +40,8 @@ public class InstallWizardActivity extends AppCompatActivity {
     private static final String BUNDLE_CSAH = "csah";
     private static final String BUNDLE_BACKUPUTILS = "backuputils";
 
+    private static final boolean DO_LEGACY_WRITE_STORAGE = android.os.Build.VERSION.SDK_INT <= android.os.Build.VERSION_CODES.Q; // <= SDK29
+
     public enum WizardMode {
         WIZARDMODE_DEFAULT(0),
         WIZARDMODE_RETURNING(1),
@@ -54,7 +56,7 @@ public class InstallWizardActivity extends AppCompatActivity {
 
     private enum WizardStep {
         WIZARD_START,
-        WIZARD_PERMISSIONS, WIZARD_PERMISSIONS_LOCATION,
+        WIZARD_PERMISSIONS, WIZARD_PERMISSIONS_LOCATION, WIZARD_PERMISSIONS_LEGACY_WRITE_STORAGE,
         WIZARD_PERMISSIONS_BASEFOLDER, WIZARD_PERMISSIONS_MAPFOLDER, WIZARD_PERMISSIONS_MAPTHEMEFOLDER, WIZARD_PERMISSIONS_GPXFOLDER, WIZARD_PERMISSIONS_BROUTERTILESFOLDER,
         WIZARD_PLATFORMS,
         WIZARD_ADVANCED,
@@ -70,6 +72,7 @@ public class InstallWizardActivity extends AppCompatActivity {
     private static final int REQUEST_CODE_WIZARD_GC = 0x7167;
 
     private final PermissionAction<Void> askLocationPermissionAction = PermissionAction.register(this, PermissionContext.LOCATION, b -> gotoNext());
+    private final PermissionAction<Void> askLegacyStoragePermissionAction = PermissionAction.register(this, PermissionContext.LEGACY_WRITE_EXTERNAL_STORAGE, b -> gotoNext());
 
     // dialog elements
     private ImageView logo = null;
@@ -173,8 +176,12 @@ public class InstallWizardActivity extends AppCompatActivity {
             case WIZARD_PERMISSIONS_LOCATION:
                 title.setText(PermissionContext.LOCATION.getExplanationTitle());
                 PermissionContext.LOCATION.getExplanation().applyTo(text);
-                //text.setText(R.string.location_permission_request_explanation);
                 setNavigation(this::gotoPrevious, 0, null, 0, this::requestLocation, 0);
+                break;
+            case WIZARD_PERMISSIONS_LEGACY_WRITE_STORAGE:
+                title.setText(PermissionContext.LEGACY_WRITE_EXTERNAL_STORAGE.getExplanationTitle());
+                PermissionContext.LEGACY_WRITE_EXTERNAL_STORAGE.getExplanation().applyTo(text);
+                setNavigation(this::gotoPrevious, 0, null, 0, this::requestLegacyWriteStorage, 0);
                 break;
             case WIZARD_PERMISSIONS_BASEFOLDER:
                 setFolderInfo(PersistableFolder.BASE, R.string.wizard_basefolder_request_explanation, false);
@@ -240,6 +247,8 @@ public class InstallWizardActivity extends AppCompatActivity {
                 final StringBuilder info = new StringBuilder();
                 info.append(getString(R.string.wizard_status_title)).append(":\n")
                         .append(getString(R.string.permission_location_explanation_title)).append(": ").append(hasLocationPermission() ? getString(android.R.string.ok) : getString(R.string.status_not_ok)).append("\n")
+                        .append(DO_LEGACY_WRITE_STORAGE ?
+                                getString(R.string.permission_legacy_write_external_storage_explanation_title) + ": " + (hasLegacyWriteStoragePermission() ? getString(android.R.string.ok) : getString(R.string.status_not_ok)) + "\n" : "")
                         .append(getString(R.string.wizard_status_basefolder)).append(": ").append(ContentStorageActivityHelper.baseFolderIsSet() ? getString(android.R.string.ok) : getString(R.string.status_not_ok)).append("\n")
                         .append(getString(R.string.wizard_status_platform));
                 boolean platformConfigured = false;
@@ -358,7 +367,8 @@ public class InstallWizardActivity extends AppCompatActivity {
     }
 
     private boolean stepCanBeSkipped() {
-        return (step == WizardStep.WIZARD_PERMISSIONS && (Build.VERSION.SDK_INT < Build.VERSION_CODES.M || hasLocationPermission()))
+        return (step == WizardStep.WIZARD_PERMISSIONS && (Build.VERSION.SDK_INT < Build.VERSION_CODES.M || (hasLocationPermission() && (!DO_LEGACY_WRITE_STORAGE || hasLegacyWriteStoragePermission()))))
+                || (step == WizardStep.WIZARD_PERMISSIONS_LEGACY_WRITE_STORAGE && (Build.VERSION.SDK_INT < Build.VERSION_CODES.M || !DO_LEGACY_WRITE_STORAGE || hasLegacyWriteStoragePermission()))
                 || (step == WizardStep.WIZARD_PERMISSIONS_LOCATION && (Build.VERSION.SDK_INT < Build.VERSION_CODES.M || hasLocationPermission()))
                 || (step == WizardStep.WIZARD_PERMISSIONS_BASEFOLDER && ContentStorageActivityHelper.baseFolderIsSet())
                 || (step == WizardStep.WIZARD_PERMISSIONS_MAPFOLDER && !mapFolderNeedsMigration())
@@ -386,7 +396,7 @@ public class InstallWizardActivity extends AppCompatActivity {
 
     public static boolean isConfigurationOk() {
         final boolean isPlatformConfigured = ConnectorFactory.getActiveConnectorsWithValidCredentials().length > 0;
-        return hasLocationPermission() && isPlatformConfigured && ContentStorageActivityHelper.baseFolderIsSet();
+        return hasLocationPermission() && hasLegacyWriteStoragePermission() && isPlatformConfigured && ContentStorageActivityHelper.baseFolderIsSet();
     }
 
     public static boolean needsFolderMigration() {
@@ -403,6 +413,15 @@ public class InstallWizardActivity extends AppCompatActivity {
     private void requestLocation() {
         setSkip(this::gotoNext, 0);
         askLocationPermissionAction.launch(null, true);
+    }
+
+    private void requestLegacyWriteStorage() {
+        setSkip(this::gotoNext, 0);
+        askLegacyStoragePermissionAction.launch(null, true);
+    }
+
+    private static boolean hasLegacyWriteStoragePermission() {
+        return !DO_LEGACY_WRITE_STORAGE || PermissionContext.LEGACY_WRITE_EXTERNAL_STORAGE.hasAllPermissions();
     }
 
 

--- a/main/src/main/java/cgeo/geocaching/permission/PermissionContext.java
+++ b/main/src/main/java/cgeo/geocaching/permission/PermissionContext.java
@@ -22,9 +22,8 @@ public enum PermissionContext {
 
     LOCATION(new String[]{Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION}, R.string.permission_location_explanation, R.string.permission_location_explanation_title),
     SEARCH_USER_IN_CONTACTS(new String[]{Manifest.permission.READ_CONTACTS}, R.string.permission_contacts_read_explanation, R.string.permission_contacts_read_explanation_title),
-    SHOW_WALLPAPER(new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, R.string.permission_read_external_storage_explanation, R.string.permission_read_external_storage_explanation_title);
-
-
+    SHOW_WALLPAPER(new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, R.string.permission_read_external_storage_explanation, R.string.permission_read_external_storage_explanation_title),
+    LEGACY_WRITE_EXTERNAL_STORAGE(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, R.string.permission_legacy_write_external_storage_explanation, R.string.permission_legacy_write_external_storage_explanation_title);
 
     private final String[] permissions;
     @StringRes

--- a/main/src/main/res/values/strings.xml
+++ b/main/src/main/res/values/strings.xml
@@ -2452,6 +2452,9 @@
     <string name="permission_location_explanation">c:geo is optimized for usage with GPS data (fine location) to locate your position and calculate distance and direction to geocaches. Without this permission, location-based functions will not work as expected.</string>
     <string name="permission_read_external_storage_explanation_title">Read external Storage permission</string>
     <string name="permission_read_external_storage_explanation">c:geo can use your systems wallpaper as background for the app. To do this, c:geo needs permission to read your media data on external storage.</string>
+    <string name="permission_legacy_write_external_storage_explanation">c:geo will write data onto your phone storage or SD card e.g. when you save geocaches for offline use, import or export files and other functions. On some Android versions, you need to grant an additional permission for these actions.</string>
+    <string name="permission_legacy_write_external_storage_explanation_title">Legacy Write External Storage permission</string>
+
 
     <string name="storage_permission_needed">This function is not possible without storage permission.</string>
 

